### PR TITLE
Glue Launcher Lambda, Tests & Supports

### PR DESCRIPTION
**Pull request by connoravo**
Glue launcher lambda with associated unit tests
Supporting GHA

Signed-off-by: Connor Avery <ConnorAvery@digital.uc.dwp.gov.uk>

### CODE REVIEW

File: tests/test_glue_launcher.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+args.mismatched_timestamps_table_name = "mismatched_timestamps"
+args.manifest_s3_input_parquet_location_missing_import = "/missing_imports"
+args.manifest_s3_input_parquet_location_missing_export = "/missing_exports"
+args.manifest_s3_input_parquet_location_counts = "/counts"
+args.manifest_s3_input_parquet_location_mismatched_timestamps = "/mismatched_timestamps"
+args.manifest_s3_output_location = "s3://bucket/output_location"
+
+
+class TestRetriever(unittest.TestCase):
+    # @mock.patch(
```
Comment (steveburton4) [inline tests/test_glue_launcher.py:52]: This needs removing or uncommenting

File: src/glue_launcher_lambda/glue_launcher.py [hunk 2]
```diff
@@ -0,0 +1,517 @@
...
+    return boto3.service("glue")
+
+
+def check_running_batch_tasks(job_queue, batch_client):
+    """Check the AWS Batch job queue, return count of tasks in each status"""
+
+    operational_tasks = 0
+
+    for status in OPERATIONAL_JOB_STATUSES:
+        response = batch_client.list_jobs(
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:259]: Does this call need paging?

File: src/glue_launcher_lambda/glue_launcher.py [hunk 3]
```diff
@@ -0,0 +1,517 @@
...
+    dumped_event = get_escaped_json_string(event)
+    logger.info(f'Event", "event": {dumped_event}, "mode": "handler')
+
+    detail_dict = get_and_validate_job_details(event)
+
+    job_name = detail_dict[JOB_NAME_KEY]
+    job_status = detail_dict[JOB_STATUS_KEY]
+    job_queue = detail_dict[JOB_QUEUE_KEY]
+
+    if job_status in IGNORED_JOB_STATUSES:
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:468]: This is vulnerable to another status being added by batch. I would prefer to do something like `if job_status not in FINISHED_JOB_STATUSES` or something.

File: src/glue_launcher_lambda/glue_launcher.py [hunk 4]
```diff
@@ -0,0 +1,517 @@
...
+        ]
+    else:
+        _args.manifest_comparison_cut_off_date_start = "1983-11-15T09:09:55.000"
+
+    if "MANIFEST_COMPARISON_CUT_OFF_DATE_END" in os.environ:
+        _args.manifest_comparison_cut_off_date_end = os.environ[
+            "MANIFEST_COMPARISON_CUT_OFF_DATE_END"
+        ]
+    else:
+        _args.manifest_comparison_cut_off_date_end = "2099-11-15T09:09:55.000"
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:145]: I'm not sure how you will be able to pass this and the above in from the terraform as they need to be relevant to the time running. I would set the default here to the default that is needed when run. Concourse usually sets them in bash and passes them at runtime, but that isn't possible here, so they need to be set at runtime here.

Cut off date end should be midnight of current day and cut off date start should default to midnight of yesterday I think. Worth checking the manifest comparison concourse code to be 100% sure.

File: src/glue_launcher_lambda/glue_launcher.py [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+
+    if "MANIFEST_COMPARISON_MARGIN_OF_ERROR_MINUTES" in os.environ:
+        _args.manifest_comparison_margin_of_error_minutes = os.environ[
+            "MANIFEST_COMPARISON_MARGIN_OF_ERROR_MINUTES"
+        ]
+    else:
+        _args.manifest_comparison_margin_of_error_minutes = "2"
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:152]: I think it defaults to 0 in manifest comparison pipeline in concourse, check that and copy here.

File: src/glue_launcher_lambda/glue_launcher.py [hunk 6]
```diff
@@ -0,0 +1,517 @@
...
+    """
+    timestamp = datetime.strptime(formatted_timestamp_string, "%Y-%m-%dT%H:%M:%S.%f")
+    if minutes_to_add > 0:
+        timestamp = timestamp + timedelta(minutes=minutes_to_add)
+
+    return int((timestamp - datetime(1970, 1, 1)).total_seconds() * 1000.0)
+
+
+def get_batch_client():
+    return boto3.client("batch")
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:242]: Add retries to all these clients. See the ADG python code for the generate_dataset step. I added retries in there, you can copy that.

File: src/glue_launcher_lambda/glue_launcher.py [hunk 7]
```diff
@@ -0,0 +1,517 @@
...
+            console_printer.print_info(
+                f"Athena query execution finished in {str(time_taken)} seconds with status of '{state}'"
+            )
+            return state
+
+        time.sleep(1)
+        time_taken += 1
+
+
+def execute_athena_query(output_location, query, athena_client):
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:340]: I can't see a test for this method that ensures the client itself has the right calls

File: src/glue_launcher_lambda/glue_launcher.py [hunk 8]
```diff
@@ -0,0 +1,517 @@
...
+    return tables
+
+
+def fetch_table_drop_sql_file(file_path, args=None):
+    with open(os.path.join(file_path, "drop-table.sql"), "r") as f:
+        base_drop_query = f.read()
+        return base_drop_query
+
+
+def poll_athena_query_status(id):
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:318]: I can't see a test for this method that ensures the client itself has the right calls

Comment (steveburton4) [reply src/glue_launcher_lambda/glue_launcher.py:318]: Also we should pass the athena client in here

File: src/glue_launcher_lambda/glue_launcher.py [hunk 9]
```diff
@@ -0,0 +1,517 @@
...
+
+def get_athena_client():
+    return boto3.client("athena")
+
+
+def get_glue_client():
+    return boto3.service("glue")
+
+
+def check_running_batch_tasks(job_queue, batch_client):
```
Comment (steveburton4) [inline src/glue_launcher_lambda/glue_launcher.py:253]: I can't see a test for this method that ensures the client itself has the right calls

Comment (steveburton4) [reply src/glue_launcher_lambda/glue_launcher.py:253]: Just mock the client itself, easy when passed in with mock.patch

File: src/glue_launcher_lambda/glue_launcher.py (hunk 6 above)
Comment (steveburton4) [reply src/glue_launcher_lambda/glue_launcher.py:242]: Example -> https://github.com/dwp/aws-analytical-dataset-generation/blob/master/steps/generate_dataset_from_htme.py#L625

File: src/glue_launcher_lambda/glue_launcher.py (hunk 2 above)
Comment (steveburton4) [reply src/glue_launcher_lambda/glue_launcher.py:259]: I'm sure there is an example for this somewhere, but I can't find it. Search private git for `nextToken` and I'm sure there is an example!

Comment (steveburton4) [reply src/glue_launcher_lambda/glue_launcher.py:259]: You basically do one call and retrieve the `NextToken` from the result. And then you do a loop _while next token is null_ to get the result, refreshing the next token from the latest result each time. See https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/batch.html#Batch.Client.list_jobs

File: src/glue_launcher_lambda/glue_launcher.py (hunk 5 above)
Comment (connoravo) [reply src/glue_launcher_lambda/glue_launcher.py:152]: I checked, 2 minutes is what it is set as in the behave framework.

**Review by steveburton4 — APPROVED** [review]

**Pull request merged**